src/www/app/server/access_config.ts

```diff
@@ -0,0 +1,175 @@
+// Copyright 2018 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {makeConfig, SHADOWSOCKS_URI, SIP002_URI} from 'ShadowsocksConfig';
+import {ServerUrlInvalid, ServerIncompatible, ShadowsocksUnsupportedCipher} from '../../model/errors';
+import {ShadowsocksConfig} from '../config';
+
+export enum OutlineServerSupportedCipher {
+  CHACHA20_IETF_POLY1305 = 'chacha20-ietf-poly1305',
+  AES_128_GCM = 'aes-128-gcm',
+  AES_192_GCM = 'aes-192-gcm',
+  AES_256_GCM = 'aes-256-gcm',
+}
+
+export enum OutlineServerAccessType {
+  SHADOWSOCKS_URI,
+  SHADOWSOCKS_CONFIG,
+}
+
+export class OutlineServerAccessConfig implements ShadowsocksConfig {
+  type: OutlineServerAccessType;
+  isOutlineServer = false;
+
+  private _rawUri?: string;
+  private _host?: string;
+  private _port?: number;
+  private _method?: OutlineServerSupportedCipher;
+  private _password?: string;
+  private _name?: string;
+
+  constructor(accessData: string | ShadowsocksConfig) {
+    if (typeof accessData === 'string') {
+      this.fromUri(accessData);
+    } else {
+      this.fromShadowsocksConfig(accessData);
+    }
+  }
+
+  get host() {
+    return this._host;
+  }
+
+  get port() {
+    return this._port;
+  }
+
+  get address() {
+    if (!this._host) {
+      return 'server-address-pending';
```

Comment:
This might confuse the caller because it will either return a raw string or a localized message ID.

Reply:
yeah i'm not sure how to do this. i guess just return nothing in this case

Reply:
which means the server model adress needs be optional as well

Reply:
Plus one. We shouldn't return different types.

Reply:
Yeah, i had already started reverting this locally